### PR TITLE
Download ISO to temp file before writing to USB stick

### PR DESCRIPTION
**Pull request by GBA-17**
Previously, the `download()` function would create a file on the USB stick and write to the file during the download. This caused problems when the download was interrupted, since we would end up with a corrupted ISO file in the cache.

Instead of writing the download directly to the cache, this PR first writes the download to a temp file. The temp file is always removed at the end of the function.

If the download succeeds, the temp file is copied to the USB cache. Otherwise, if the download fails or is interrupted by the user, the temp file is just removed and the USB cache is not modified.

### CODE REVIEW

**Review by rminnich — DISMISSED** [review]
will the cancel work if it happens on the copy to the usb stick? Should there be some information printed when that starts?

**Comment by GBA-17** [comment]
That's a good point, maybe we can run a second progress screen when the copy happens?

**Comment by GBA-17** [comment]
I just added a progress screen for the copy step, but it is ok if the user can't cancel during that?

File: cmds/webboot/utils.go [hunk 1]
```diff
@@ -51,17 +53,35 @@ func download(URL, fPath string, uiEvents <-chan ui.Event) error {
...
+		return err
+	}
+
+	cacheFile, err := os.Create(fPath)
+	if err != nil {
+		return err
+	}
+	defer cacheFile.Close()
+
+	if _, err := io.Copy(cacheFile, tempFile); err != nil {
```
Comment (rminnich) [inline cmds/webboot/utils.go:84]: If this fails will it leave a partial file on the stick? If it fails should we remove the file?

File: cmds/webboot/utils.go [hunk 2]
```diff
@@ -51,17 +53,35 @@ func download(URL, fPath string, uiEvents <-chan ui.Event) error {
...
+
+	counter.Close()
+	copyProgress := menu.NewProgress("Download complete. Writing ISO to cache", true)
+	defer copyProgress.Close()
+
+	if _, err = tempFile.Seek(0, 0); err != nil {
+		return err
+	}
+
+	cacheFile, err := os.Create(fPath)
```
Comment (rminnich) [inline cmds/webboot/utils.go:78]: Should we check to see if the file is there? if it is there, what should we do?

Comment (GBA-17) [reply cmds/webboot/utils.go:78]: Yeah, if the requested file name matches a file we already have, we can probably avoid downloading it.

Maybe we can check if the file already exists at the beginning of the function?

File: cmds/webboot/utils.go (hunk 1 above)
Comment (GBA-17) [reply cmds/webboot/utils.go:84]: Right now it would leave a partial file on disk. Maybe I can add a line to the error case to remove the file?

Comment (rminnich) [reply cmds/webboot/utils.go:84]: yes sounds good

File: cmds/webboot/utils.go (hunk 2 above)
Comment (rminnich) [reply cmds/webboot/utils.go:78]: what if it is corrupt? If the system was reboot in the middle of copying to the usb, half a file could be there. Should we give them the option of keeping it or not?

Comment (rminnich) [reply cmds/webboot/utils.go:78]: or could we do an md5sum and if it fails, remove it?

Comment (GBA-17) [reply cmds/webboot/utils.go:78]: I think the md5sum is a good idea.

Should we remove the existing ISO by default, or maybe show a confirmation screen?

Comment (rminnich) [reply cmds/webboot/utils.go:78]: I would confirm each step ....
let's see how that looks.

File: cmds/webboot/utils.go (hunk 1 above)
Comment (GBA-17) [reply cmds/webboot/utils.go:84]: I was thinking of adding `os.Remove(cacheFile.Name())` inside the error case, but would that cause any problems with the deferred `cacheFile.Close()`?

Comment (rminnich) [reply cmds/webboot/utils.go:84]: no. In fact let's talk about why tomorrow, the reason is interesting.

Comment (GBA-17) [reply cmds/webboot/utils.go:84]: Sounds good, I added the `os.Remove()` line.

File: cmds/webboot/utils.go (hunk 2 above)
Comment (GBA-17) [reply cmds/webboot/utils.go:78]: I think we can do the file exists checks outside of the download function. I'll put up a PR for that in a bit.

**Comment by GBA-17** [comment]
I think this PR is mostly ready to go.

The only thing is that we don't check whether the file already exists, but I think we can do that outside of the download function.

**Review by pallaud — DISMISSED** [review]
Approving this, and noting that the checks for whether a file exists will be added in a follow up PR.

**Review by pallaud — APPROVED** [review]

**Pull request merged**